c++/src/io/Cache.hh

```diff
@@ -0,0 +1,129 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "orc/MemoryPool.hh"
+#include "orc/OrcFile.hh"
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <future>
+#include <utility>
+#include <vector>
+
+namespace orc {
+
+  struct ReadRange {
+    uint64_t offset;
+    uint64_t length;
+
+    ReadRange() = default;
+    ReadRange(uint64_t offset, uint64_t length) : offset(offset), length(length) {}
+
+    friend bool operator==(const ReadRange& left, const ReadRange& right) {
+      return (left.offset == right.offset && left.length == right.length);
+    }
+    friend bool operator!=(const ReadRange& left, const ReadRange& right) {
+      return !(left == right);
+    }
+
+    bool contains(const ReadRange& other) const {
+      return (offset <= other.offset && offset + length >= other.offset + other.length);
+    }
+  };
+
+  struct ReadRangeCombiner {
+    const uint64_t holeSizeLimit;
+    const uint64_t rangeSizeLimit;
+
+    std::vector<ReadRange> coalesce(std::vector<ReadRange> ranges) const;
+
+    static std::vector<ReadRange> coalesceReadRanges(std::vector<ReadRange> ranges,
+                                                     uint64_t holeSizeLimit,
+                                                     uint64_t rangeSizeLimit);
+  };
+
+  using Buffer = DataBuffer<char>;
+  using BufferPtr = std::shared_ptr<Buffer>;
+
+  struct RangeCacheEntry {
+    ReadRange range;
+
+    // The result may be get multiple times, so we use shared_future instead of std::future
+    BufferPtr buffer;
+    std::shared_future<void> future;
```

Comment:
```suggestion
    ReadRange range;
    BufferPtr buffer;
    std::shared_future<void> future;  // use shared_future in case of multiple get calls
```